neo/Consensus/ConsensusService.cs

```diff
@@ -290,16 +290,10 @@ private void OnPrepareResponseReceived(ConsensusPayload payload, PrepareResponse
         {
             Log($"{nameof(OnPrepareResponseReceived)}: height={payload.BlockIndex} view={message.ViewNumber} index={payload.ValidatorIndex}");
             if (context.Signatures[payload.ValidatorIndex] != null) return;
+            context.Signatures[payload.ValidatorIndex] = message.Signature;
```

Comment:
This is wrong. Because the signature is only assigned when `if` and `else if`, not for all the case.

Reply:
I understand it in the sense of a blacklist of the CN that sent this null payload.

However, I think that we would need some additional verification to be sure that the node that really sent this and that it was not another one trying to blacklist/lie about that node.

Reply:
https://github.com/neo-project/neo/blob/c32c63b1b623ff0a2cb4839292ab67bc881223f7/neo/Consensus/ConsensusService.cs#L254-L257

Reply:
Yes, but these lines are only in the case that `PrepareRequest` arrived after `PrepareResponse`, which is not a normal operation.

The case I mention is (with 7 nodes) :
* A (primary) sends  `PrepareRequest` to `B`
* `B` receives `PrepareResponse` from `C` and `D`, but both are null (wrong payloads)
* `B` receives a correct   `PrepareResponse` from `E and then goes to `CheckSignatures` and send relay the Block

Reply:
Before `PrepareRequest`:

https://github.com/neo-project/neo/blob/c32c63b1b623ff0a2cb4839292ab67bc881223f7/neo/Consensus/ConsensusService.cs#L254-L257

After `PrepareRequest`:

https://github.com/neo-project/neo/blob/c32c63b1b623ff0a2cb4839292ab67bc881223f7/neo/Consensus/ConsensusService.cs#L298

Reply:
@erikzhang, in addition, in the #426 all messages carries the `PrepareRequest` Payload. Thus, for that case in which `PrepareResponse` arrives prior to `PrepareRequest` we can easily run and verify.

Thus, the best route is ensured for each node! :dagger:

Reply:
ERIKKKKKKKKK, I think that I finally understood you.

You are not checking the `ConsensusPayload payload`! You are checking it against your local data (the only case that will force it to be empty).
Sorry about that =/ You are right again.

But, then, I will reformulate my case to:
* `B` never receives a PrepareRequest (lost forever)
* `B` receives `PrepareResponse` from `C`, `D`, `E` and `F` , four of them are null (wrong payloads)
* `B` receives a correct  `PrepareResponse` from `G` and then goes to `CheckSignatures` relaying the Block with signatures: 
      - `G` is the only correct signatures
      -  `C`, `D`, `E` and `F` incorrect (not validated because payload was null)

Reply:
https://github.com/neo-project/neo/blob/c32c63b1b623ff0a2cb4839292ab67bc881223f7/neo/Consensus/ConsensusService.cs#L294-L297

If B never receives a PrepareRequest (lost forever), it won't run `CheckSignatures`.

Reply:
I had edited my comment after understanding that, It won't run CheckSignatures...Sorry for that =/

Reply:
I updated the new commit phase PR considering this new explanation.

https://github.com/neo-project/neo/pull/426/commits/d3cc47970d680580510ae7dff0e1c0895676410c